### PR TITLE
Priority queue based stakeManager

### PR DESCRIPTION
**Pull request by 0xAshish**

### CODE REVIEW

**Comment by ashishrp** [comment]
WIP: please keep this PR open.

File: contracts/root/StakeManager.sol [hunk 1]
```diff
@@ -6,182 +6,244 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
-  bytes32 public epochSeed = keccak256(abi.encodePacked(block.difficulty + block.number + now));
+  //Todo: dynamically update
+  uint256 public minStakeAmount = 0;
+  uint256 public minLockInPeriod = 1; //(unit epochs)
+  uint256 public stakingIdCount = 0;  // just a counter/index to map it with PQ w/address
 
-  //
-  // Events
-  //
+  enum State { isStaker, isValidator, isProposer, isBaned, isUnstaking }
```
Comment (jdkanani) [inline contracts/root/StakeManager.sol]: Change to `ValidatorStatus`. Status can be any one of these - `WAITING`, `SET`, & `LEAVING`, `LEFT` or something similar.

File: contracts/root/StakeManager.sol [hunk 2]
```diff
@@ -6,182 +6,244 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
+  address[] public currentValidators;
 
-  //
-  // Constructor
-  //
+  mapping (address=>Staker) stakers; 
+  mapping (uint256=>address) stakingIdToAddress;
 
-  constructor (address _token) public {
+  constructor(address _token) public payable {
```
Comment (jdkanani) [inline contracts/root/StakeManager.sol]: Why constructor is payable?

File: contracts/root/StakeManager.sol [hunk 3]
```diff
@@ -6,182 +6,244 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
     require(totalStakedFor(msg.sender) > 0);
     _;
   }
 
-  //
-  // ERC900 implementation
-  //
+  
+  function _priority(address user, uint256 amount, bytes data)internal view returns(uint256){
+    // priority = priority << 64 | amount.div(totalStake) ;
```
Comment (jdkanani) [inline contracts/root/StakeManager.sol]: Please explain comments

File: contracts/root/StakeManager.sol [hunk 4]
```diff
@@ -6,182 +6,244 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
+    // return amount.mul(10000000).add(currentEpoch.mul(1000).add(amount.mul(100).div(user.balance)));
+    return amount;
+  }
 
   function stake(uint256 amount, bytes data) public {
-    stakeFor(msg.sender, amount, data);
+    // add condition to update already existing users stake !? 
+    if (stakers[msg.sender].epoch==0) {
```
Comment (jdkanani) [inline contracts/root/StakeManager.sol]: Lint, lint and lint

Comment (jdkanani) [reply contracts/root/StakeManager.sol]: Read more code

File: contracts/root/StakeManager.sol (hunk 3 above)
Comment (jdkanani) [reply contracts/root/StakeManager.sol]: Shouldn't function be `pure` instead of view?

File: contracts/root/StakeManager.sol [hunk 5]
```diff
@@ -6,182 +6,244 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
     totalStake = totalStake.add(amount);
-
-    // broadcast event
-    emit Staked(user, amount, stakedAmount, data);
+    emit Staked(user, amount, totalStake, data);
   }
+  
+  // returns validators
+  function getValidators() public returns (address[]) {
+    // add condition for currentSize of PQ
```
Comment (jdkanani) [inline contracts/root/StakeManager.sol]: Get validators should be `view`. Make necessary changes for the same.

Comment (jdkanani) [reply contracts/root/StakeManager.sol]: No modifiers? Any can call this method?

File: contracts/root/StakeManager.sol (hunk 3 above)
Comment (ashishrp) [reply contracts/root/StakeManager.sol]: priority calculation logic is commented, need to test it's correctness. After that will update priority function.

File: contracts/root/StakeManager.sol (hunk 5 above)
Comment (ashishrp) [reply contracts/root/StakeManager.sol]: it is updating the state of contract.

File: contracts/root/StakeManager.sol [hunk 6]
```diff
@@ -6,134 +6,194 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
-      stakeList.push(Stake({
-        amount: 0,
-        staker: user
-      }));
-      stakers[user] = stakeList.length - 1;
+    require(amount >= minStakeAmount); 
+
+    // actual staker cannot be on index 0
+    if (stakersList.length == 0) {
+      stakersList.push(address(0x0));
```
Comment (jdkanani) [inline contracts/root/StakeManager.sol]: Do we need this now?

File: contracts/root/StakeManager.sol [hunk 7]
```diff
@@ -6,134 +6,194 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
+      stakers[address(0x0)] = Staker(0, 0, new bytes(0), ValidatorStatus.WAITING, stakingIdCount);
```
Comment (jdkanani) [inline contracts/root/StakeManager.sol]: Let's use proper struct instead to make it more readable:

```
stakers[address(0x0)] = Staker({
   epoch: 0,
   amount: 0,
   bytes data: new bytes(0x0),
   status: ValidatorStatus.WAITING,
   stakingId: stakingIdCount
})
```

File: contracts/root/StakeManager.sol [hunk 8]
```diff
@@ -6,134 +6,194 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
+      stakingIdToAddress[stakingIdCount] = address(0x0);
```
Comment (jdkanani) [inline contracts/root/StakeManager.sol]: Do we need stakingIdToAddress?

File: contracts/root/StakeManager.sol [hunk 9]
```diff
@@ -6,134 +6,194 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
-
+    // transfer tokens to stake manager
+    require(tokenObj.transferFrom(user, address(this), amount));
+    
     // update total stake
     totalStake = totalStake.add(amount);
 
-    // broadcast event
-    emit Staked(user, amount, stakedAmount, data);
+    stakers[user] = Staker(currentEpoch, amount, data, ValidatorStatus.WAITING, stakingIdCount);
```
Comment (jdkanani) [inline contracts/root/StakeManager.sol]: Same as above. Make obj creation more readable

File: contracts/root/StakeManager.sol [hunk 10]
```diff
@@ -6,134 +6,194 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
+        // Todo: delete from staker list
+        // stakersList[] = stakersList[] delete index 
+      }
+    }
+  }
 
-    // transfer stake amount
-    tokenObj.transfer(msg.sender, amount);
+  function unstake(uint256 amount, bytes data) public onlyStaker { // onlyownder
+    // require(stakers[msg.sender]); //staker exists
```
Comment (jdkanani) [inline contracts/root/StakeManager.sol]: Remove comment

File: contracts/root/ValidatorSet.sol [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+  function addValidator(address validator, uint256 votingPower) public {
+    require(votingPower > 0);
+    validators.push(Validator(votingPower, 0, validator)); //use index instead
+    if(lowestPower > votingPower ) { 
+      lowestPower = votingPower;
+    }
+    totalVotingPower += votingPower;
+  }
+
+  function _getProposer() public returns(address) {
```
Comment (jdkanani) [inline contracts/root/ValidatorSet.sol]: Let's not use `_` when function is public.

File: contracts/root/StakeManager.sol [hunk 12]
```diff
@@ -6,134 +6,194 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
+    require(stakers[msg.sender].epoch != 0); 
+    require(stakers[msg.sender].amount == amount);
+    
+    stakers[msg.sender].status = ValidatorStatus.UNSTAKING;
+    exiterList.push(msg.sender); 
+    
+    emit UnstakeInit(msg.sender, amount, totalStake.sub(amount), "0");
+  }
+  
+  function selectProposer() public onlyRootChain returns(address) {
```
Comment (jdkanani) [inline contracts/root/StakeManager.sol]: Can we change it to `getProposer()` and add `view` as it's public and just getter (not changing contract state)?

File: contracts/root/ValidatorSet.sol (hunk 11 above)
Comment (jdkanani) [reply contracts/root/ValidatorSet.sol]: can we change the name if it is changing the state? `changeProposer`/`selectProposer` or something?

File: contracts/root/StakeManager.sol [hunk 13]
```diff
@@ -6,134 +6,194 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
+      ( , stakerId) = stakeQueue.delMin();
+      validator = stakingIdToAddress[stakerId];
+      stakers[validator].status = ValidatorStatus.VALIDATOR;
+      validatorSet.addValidator(validator, stakers[validator].amount);
+      delete stakingIdToAddress[stakerId];
+    }
+    totalEpoch = currentEpoch.add(ValidatorSet.totalPower.div(
+      ValidatorSet.lowestPower).mul(_validatorThreshold));
+    
+    emit NewValidatorSet(_validatorThreshold, validatorSet.totalVotingPower, "0");
```
Comment (jdkanani) [inline contracts/root/StakeManager.sol]: Use `0x0` instead of `"0"`

File: contracts/root/StakeManager.sol [hunk 14]
```diff
@@ -6,134 +6,194 @@ import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
...
+  function _unstake() private {
+    for (uint256 i = 0; i < exiterList.length; i++) {
+      address exiter = exiterList[i];
+      
+      if (stakers[exiter].status == ValidatorStatus.UNSTAKING && (
+        currentEpoch.sub(stakers[exiter].epoch)) <= minLockInPeriod ) {
+          
+        require(tokenObj.transfer(exiter, stakers[exiter].amount));
+        totalStake = totalStake.sub(stakers[exiter].amount);
+        emit Unstaked(exiter, stakers[exiter].amount, totalStake, "0");
```
Comment (jdkanani) [inline contracts/root/StakeManager.sol]: Use `0x0` instead of `"0"`

File: contracts/root/StakeManager.sol (hunk 12 above)
Comment (ashishrp) [reply contracts/root/StakeManager.sol]: It's changing the state via validatorset contract.

File: contracts/root/ValidatorSet.sol (hunk 11 above)
Comment (ashishrp) [reply contracts/root/ValidatorSet.sol]: changed it to selectProposer, and merged both functions to save on gas.

File: contracts/root/StakeManager.sol (hunk 8 above)
Comment (ashishrp) [reply contracts/root/StakeManager.sol]: yes it's for PQ's uint to address mapping, maybe will modify PQ to support address or use Balanced BST to improve on some of the things.

**Comment by jdkanani** [comment]
Closing for now.

**Pull request closed without merging**